conf/schemas.go

```diff
@@ -58,50 +91,114 @@ func ReadSchemas(file string) (Schemas, error) {
 		}
 
 		if sec.ValueOf("pattern") == "" {
-			return nil, fmt.Errorf("[%s]: empty pattern", schema.Name)
+			return Schemas{}, fmt.Errorf("[%s]: empty pattern", schema.Name)
 		}
 		schema.Pattern, err = regexp.Compile(sec.ValueOf("pattern"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
 		}
 
 		schema.Retentions, err = ParseRetentions(sec.ValueOf("retentions"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
 		}
 
 		p := int64(0)
 		if sec.ValueOf("priority") != "" {
 			p, err = strconv.ParseInt(sec.ValueOf("priority"), 10, 0)
 			if err != nil {
-				return nil, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
+				return Schemas{}, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
 			}
 		}
 		schema.Priority = int64(p)<<32 - int64(i) // to sort records with same priority by position in file
 
 		schemas = append(schemas, schema)
 	}
 
-	sort.Sort(schemas)
-	return schemas, nil
+	sort.Sort(SchemaSlice(schemas))
+
+	return NewSchemas(schemas), nil
 }
 
 // Match returns the correct schema setting for the given metric
 // it can always find a valid setting, because there's a default catch all
 // also returns the index of the setting, to efficiently reference it
-func (s Schemas) Match(metric string) (uint16, Schema) {
-	for i, schema := range s {
+func (s Schemas) Match(metric string, interval int) (uint16, Schema) {
```

Comment:
yes, this function is complex. It took a long time to come to this solution that only requires a single call to matchString(), but still allows a single schema ID to be passed around as a reference.

I think @replay's illustration will go a long way to clarifying what is going on.  But i would change it to

![image](https://cloud.githubusercontent.com/assets/8415072/24691300/3e6d8aee-1a05-11e7-93c5-bdf623fb2a6f.png)